batch/sql/estimated-current.sql

```diff
@@ -1306,6 +1338,85 @@ BEGIN
+    SELECT user, inst_coll, 0,
+      -1 * (@n_ready_cancellable_jobs := COALESCE(SUM(n_ready_cancellable_jobs), 0)),
+      -1 * (@ready_cancellable_cores_mcpu := COALESCE(SUM(ready_cancellable_cores_mcpu), 0)),
+      -1 * (@n_running_cancellable_jobs := COALESCE(SUM(n_running_cancellable_jobs), 0)),
+      -1 * (@running_cancellable_cores_mcpu := COALESCE(SUM(running_cancellable_cores_mcpu), 0)),
+      -1 * (@n_creating_cancellable_jobs := COALESCE(SUM(n_creating_cancellable_jobs), 0)),
+      COALESCE(SUM(n_ready_cancellable_jobs), 0),
+      COALESCE(SUM(n_running_cancellable_jobs), 0),
+      COALESCE(SUM(n_creating_cancellable_jobs), 0)
+    FROM job_group_inst_coll_cancellable_resources
+    JOIN batches ON batches.id = job_group_inst_coll_cancellable_resources.batch_id
+    INNER JOIN batch_updates ON job_group_inst_coll_cancellable_resources.batch_id = batch_updates.batch_id AND
+      job_group_inst_coll_cancellable_resources.update_id = batch_updates.update_id
+    WHERE job_group_inst_coll_cancellable_resources.batch_id = in_batch_id AND
+      job_group_inst_coll_cancellable_resources.job_group_id = in_job_group_id AND
+      batch_updates.committed
+    GROUP BY user, inst_coll
+    ON DUPLICATE KEY UPDATE
+      n_ready_jobs = n_ready_jobs - @n_ready_cancellable_jobs,
+      ready_cores_mcpu = ready_cores_mcpu - @ready_cancellable_cores_mcpu,
+      n_running_jobs = n_running_jobs - @n_running_cancellable_jobs,
+      running_cores_mcpu = running_cores_mcpu - @running_cancellable_cores_mcpu,
+      n_creating_jobs = n_creating_jobs - @n_creating_cancellable_jobs,
+      n_cancelled_ready_jobs = n_cancelled_ready_jobs + @n_ready_cancellable_jobs,
+      n_cancelled_running_jobs = n_cancelled_running_jobs + @n_running_cancellable_jobs,
+      n_cancelled_creating_jobs = n_cancelled_creating_jobs + @n_creating_cancellable_jobs;
+
+    # delete all rows that are children of this job group
+    DELETE job_group_inst_coll_cancellable_resources FROM job_group_inst_coll_cancellable_resources
+    LEFT JOIN batch_updates ON job_group_inst_coll_cancellable_resources.batch_id = batch_updates.batch_id AND
+      job_group_inst_coll_cancellable_resources.update_id = batch_updates.update_id
+    INNER JOIN job_group_self_and_ancestors ON job_group_inst_coll_cancellable_resources.batch_id = job_group_self_and_ancestors.batch_id AND
+      job_group_inst_coll_cancellable_resources.job_group_id = job_group_self_and_ancestors.job_group_id
+    WHERE job_group_inst_coll_cancellable_resources.batch_id = in_batch_id AND
+      job_group_self_and_ancestors.ancestor_id = in_job_group_id AND
+      batch_updates.committed;
+
+    INSERT INTO job_groups_cancelled
+    SELECT batch_id, job_group_id
+    FROM job_group_self_and_ancestors
+    WHERE batch_id = in_batch_id AND ancestor_id = in_job_group_id
+    ON DUPLICATE KEY UPDATE job_group_id = job_groups_cancelled.job_group_id;
```

Comment:
Oh, this is interesting. A job group can be cancelled twice.

Hmm.

I think you need to `SELECT FOR UPDATE` the rows you plan to delete from `job_group_inst_coll_cancellable_resources`. Otherwise, if you cancel a job group and its child in parallel, the parallel `INSERT ... SELECT ...` queries could doubly remove the resources for the child group. AFAICT, [`INSERT ... SELECT ...`](https://dev.mysql.com/doc/refman/8.0/en/insert-select.html) doesn't have a `FOR UPDATE` that we can use to take the lock & insert in one query.

Reply:
Interesting. Good catch!